components/coding/profileV2/AchievementComponent.tsx

```diff
@@ -1,172 +1,161 @@
-import React, { useEffect, useState } from "react";
+import React, { useState } from "react";
 import { PencilAltIcon } from "@heroicons/react/outline";
-import { User } from "../../../graphql/fetchUserProfile";
-import { transformUserBadgeData } from "./AchievementTransformData";
-import { useQuery } from "@apollo/client";
-import { FETCH_CODING_BADGES } from "../../../graphql/coding/userBadges/fetchUserBadges";
+import { useMutation, useQuery } from "@apollo/client";
+import {
+  CodingBadge,
+  FetchBadgeResponse,
+  FETCH_CODING_BADGES,
+  IntroCourseUnit,
+} from "../../../graphql/coding/userBadges/fetchUserBadges";
+import { Button } from "../../ui/Button";
+import {
+  INSERT_USER_CODING_BADGES,
+  DELETE_USER_CODING_BADGES,
+} from "../../../graphql/coding/userBadges/updateUserCodingBadges";
+import UnitBadgeSection from "./achievementcomponenets/UnitBadgeSection";
+import handleOnSaveButtonClick from "./achievementcomponenets/handleSaveOnClick";
 
-export type BadgesSectionProps = {
-  data: User;
+export type AchievementComponentProps = {
+  userId: string;
+  isEditable: boolean;
 };
-export type userCodingBadge = {
-  id: number;
-};
-export type unitProps = {
-  unitTitle: string;
-  codingBadges: codingBadges[];
-};
-export type codingBadges = {
-  id: number;
-  title: string;
-  userCodingBadge: userCodingBadge[];
-  isAwarded: boolean;
-};
-export type unit = unitProps[];
-const AcheivementComponent = ({ user, isEditable = false }) => {
-  const { data } = useQuery(FETCH_CODING_BADGES, {
+
+const AcheivementComponent = ({
+  userId,
+  isEditable,
+}: AchievementComponentProps) => {
+  const [unitBadges, setUnitBadges] = useState<IntroCourseUnit[]>();
+  const [editMode, setEditMode] = useState(false);
+  const { data } = useQuery<FetchBadgeResponse>(FETCH_CODING_BADGES, {
+
     variables: {
-      userId: user.uid,
+      userId: userId,
+    },
+    onCompleted: (data) => {
+      setUnitBadges(data.intro_course_unit);
     },
   });
-  const [transformedData, setTransformedData] = useState([]);
-  const [editMode, setEditMode] = useState(false);
-  const handleBadgeClick = (inputBadge: codingBadges) => {
-    setTransformedData((prev) => {
-      const updatedData = prev.map((unit: unitProps) => {
-        return {
-          ...unit,
-          codingBadges: unit.codingBadges.map((badge) => {
-            if (badge.id == inputBadge.id) {
-              return {
-                ...badge,
-                isAwarded: !badge.isAwarded,
-              };
-            } else {
-              return badge;
-            }
-          }),
-        };
-      });
```

Comment:
Removed Transform function and instead used flatMap to mutate the nested list of badges into a single array of badges